src/core/snappingutils.cpp

```diff
@@ -32,11 +31,52 @@ void SnappingUtils::onMapSettingsUpdated()
   snap();
 }
 
+QgsPoint SnappingUtils::newPoint( const QgsPoint &snappedPoint, const QgsWkbTypes::Type wkbType )
+{
+  QgsPoint newPoint( QgsWkbTypes::Point, snappedPoint.x(), snappedPoint.y() );
+
+  // convert to the corresponding type for a full ZM support
+  if ( QgsWkbTypes::hasZ( wkbType ) && !QgsWkbTypes::hasM( wkbType ) )
+  {
+    newPoint.convertTo( QgsWkbTypes::PointZ );
+  }
+  else if ( !QgsWkbTypes::hasZ( wkbType ) && QgsWkbTypes::hasM( wkbType ) )
+  {
+    newPoint.convertTo( QgsWkbTypes::PointM );
+  }
+  else if ( QgsWkbTypes::hasZ( wkbType ) && QgsWkbTypes::hasM( wkbType ) )
+  {
+    newPoint.convertTo( QgsWkbTypes::PointZM );
+  }
+
+  // set z value
+  if ( QgsWkbTypes::hasZ( newPoint.wkbType() ) && QgsWkbTypes::hasZ( snappedPoint.wkbType() ) )
+  {
+    newPoint.setZ( snappedPoint.z() );
+  }
+
+  // set m value
+  if ( QgsWkbTypes::hasM( newPoint.wkbType() ) && QgsWkbTypes::hasM( snappedPoint.wkbType() ) )
+  {
+    newPoint.setM( snappedPoint.m() );
+  }
+
+  return newPoint;
+}
+
 void SnappingUtils::snap()
 {
-  QgsPointLocator::Match match = snapToMap( QPoint( mInputCoordinate.x(), mInputCoordinate.y() ) );
+  QgsPointLocator::Match match = snapToMap( QPoint( static_cast<int>( mInputCoordinate.x() ), static_cast<int>( mInputCoordinate.y() ) ) );
   mSnappingResult = SnappingResult( match );
 
+  //set point containing ZM if existing
+  QgsVectorLayer *vlayer = qobject_cast<QgsVectorLayer *>( currentLayer() );
+  if ( vlayer && match.layer() )
+  {
+    const QgsFeature ft = match.layer()->getFeature( match.featureId() );
```

Comment:
shouldn't we do this when striclty needed (i.e. when a Z/M is asked), to avoid fecthing if only X/Y is need (as they are already in the match if I'm right)